pkg/services/x509/uzi_validator.go

```diff
@@ -0,0 +1,228 @@
+		})
+		if err != nil {
+			return
+		}
+	} else if env == UziAcceptation {
+		roots, err = certsFromAssets([]string{
+			"certs/uzi-acc/test_zorg_csp_root_ca_g3.cer",
+		})
+		if err != nil {
+			return
+		}
+
+		intermediates, err = certsFromAssets([]string{
+			"certs/uzi-acc/test_uzi-register_medewerker_op_naam_ca_g3.cer",
+			"certs/uzi-acc/test_zorg_csp_level_2_persoon_ca_g3.cer",
+		})
+		if err != nil {
+			return
+		}
+	} else {
+		return nil, fmt.Errorf("unknown uzi environment: %s", env)
+	}
+
+	if crls == nil {
+		crls = NewCachedHttpCrlService()
+	}
+
+	validator = &UziValidator{
+		validator:         NewJwtX509Validator(roots, intermediates, validUziSigningAlgs(), crls),
+		contractTemplates: contractTemplates,
+	}
+	return
+}
+
+// Parse tries to parse a string as a base64 encoded JWT.
+// The jwt should contain at leas one certificate in the x509 header
+// It tries to find the contract in the given contractStore.
+// No other verifications are performed.
+// Make sure to call Verify to perform the actual crypto verifications
+func (u UziValidator) Parse(rawAuthToken string) (services.SignedToken, error) {
+	x509Token, err := u.validator.Parse(rawAuthToken)
+	if err != nil {
+		return nil, err
+	}
+	tokenField, ok := x509Token.token.Get("message")
+	if !ok {
+		return nil, fmt.Errorf("jwt did not contain token field")
+	}
+	contractText, ok := tokenField.(string)
+	if !ok {
+		return nil, fmt.Errorf("token field should contain a string")
+	}
+
+	c, err := contract.ParseContractString(contractText, *u.contractTemplates)
+	if err != nil {
+		return nil, err
+	}
+
+	return UziSignedToken{jwtX509Token: x509Token, contract: c}, nil
+}
+
+// extKeyUsageDocumentSigning is required for siging documents, according to the UZI spec.
+var extKeyUsageDocumentSigning = asn1.ObjectIdentifier{1, 3, 6, 1, 4, 1, 311, 10, 3, 12}
+
+// Verify performs all the crypto verifications like:
+// Correct hashing algorithm
+// Correct certificate tree
+// Certificates are not revoked
+// Verifies all the extra jwt fields like exp, iat and nbf.
+// Verifies if the signer attributes are valid
+func (u UziValidator) Verify(token services.SignedToken) error {
+	x509SignedToken, ok := token.(UziSignedToken)
+	if !ok {
+		return fmt.Errorf("wrong token type")
+	}
+	_, err := token.SignerAttributes()
+	if err != nil {
+		return fmt.Errorf("invalid signer attributes in uzi certificate: %w", err)
+	}
+	err = u.validator.Verify(x509SignedToken.jwtX509Token)
+	if err != nil {
+		return err
+	}
+
+	// check if the certificate has the Extended Key usage for document signing
+	keyUsageFound := false
+	for _, keyUsage := range x509SignedToken.jwtX509Token.chain[0].UnknownExtKeyUsage {
+		if keyUsage.Equal(extKeyUsageDocumentSigning) {
+			keyUsageFound = true
+			break
+		}
+	}
+	if !keyUsageFound {
+		return fmt.Errorf("certificate is missing the extended key usage for document siging (oid:1.3.6.1.4.1.311.10.3.12)")
```

Comment:
signing

Reply:
I think ObjectIdentifier has a `String()` func so you could pass `extKeyUsageDocumentSigning` in this error formatting func